tools/misc/mapping_extractor.py

```
@@ -0,0 +1,259 @@
+        df = df.drop(index=[i - 1 for i in rows_to_ignore], errors="ignore")
+
+    results = []
+    seen_pairs = set()  # To avoid duplicates
+    warnings_list = []
+
+
+    # Securely check raw Excel column headers using openpyxl
+    try:
+        wb_raw = load_workbook(filename=source_file, read_only=True, data_only=True)
+    except Exception as e:
+        raise RuntimeError(f"Unable to open Excel file with openpyxl: \"{source_file}\"\n\t\t {e}")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)